utils/s3/client.cc

```diff
@@ -42,10 +45,77 @@ future<> ignore_reply(const http::reply& rep, input_stream<char>&& in_) {
     co_await util::skip_entire_stream(in);
 }
 
+class dns_connection_factory : public http::experimental::connection_factory {
+protected:
+    std::string _host;
+    int _port;
+    struct state {
+        bool initialized = false;
+        socket_address addr;
+        ::shared_ptr<tls::certificate_credentials> creds;
+        shared_future<> done;
+
+        state(future<> f) : done(std::move(f)) {}
+    };
+    lw_shared_ptr<state> _state;
+
+    void initialize(std::optional<endpoint_config::https_config> https) {
+        promise<> initialized;
+        _state = make_lw_shared<state>(initialized.get_future());
+        when_all(
+            net::dns::get_host_by_name(_host, net::inet_address::family::INET).then([port = _port] (::net::hostent&& h) {
+                return make_ready_future<socket_address>(socket_address(h.addr_list.front(), port));
+            }),
+            [https = std::move(https)] () -> future<::shared_ptr<tls::certificate_credentials>> {
+                if (!https) {
+                    auto nil = ::shared_ptr<tls::certificate_credentials>();
+                    return make_ready_future<::shared_ptr<tls::certificate_credentials>>(nil);
+                }
+
+                return do_with(tls::credentials_builder(), *https, [] (auto& cbuild, auto& https) {
```

Comment:
nit: we can move https into do_with(). std::move(*https).

Reply:
The whole https thing is gone from here in updated version